core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -762,55 +895,141 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
 		-> Box<dyn FnMut(&PeerId, MessageIntent, &Block::Hash, &[u8]) -> bool + 'a>
 	{
 		let (inner, do_rebroadcast) = {
-			use parking_lot::RwLockWriteGuard;
-
-			let mut inner = self.inner.write();
+			let inner = Arc::clone(&self.inner);
 			let now = Instant::now();
-			let do_rebroadcast = if now >= inner.next_rebroadcast {
-				inner.next_rebroadcast = now + REBROADCAST_AFTER;
+			let do_rebroadcast = if now >= inner.read().next_rebroadcast {
+				(*inner.write()).next_rebroadcast = now + REBROADCAST_AFTER;
 				true
 			} else {
 				false
 			};
-
-			// downgrade to read-lock.
-			(RwLockWriteGuard::downgrade(inner), do_rebroadcast)
+			(inner, do_rebroadcast)
 		};
 
 		Box::new(move |who, intent, topic, mut data| {
 			if let MessageIntent::PeriodicRebroadcast = intent {
 				return do_rebroadcast;
 			}
 
-			let peer = match inner.peers.peer(who) {
-				None => return false,
-				Some(x) => x,
-			};
-
 			// if the topic is not something we're keeping at the moment,
 			// do not send.
-			let (maybe_round, set_id) = match inner.live_topics.topic_info(&topic) {
-				None => return false,
-				Some(x) => x,
+			let (maybe_round, set_id) = {
+				let inner = inner.read();
```

Comment:
It doesn't make sense to re-acquire the lock (multiple times?????) on every message. Remember that this `Box<Fn>` will be called in a loop. Would you really rather write

```rust
for message in messages {
    inner.read().should_we_send(message)?;
}
```

instead of 

```rust
let inner = inner.read();
for message in messages {
    inner.should_we_send(message)?;
}
```

Again, this is jumping the gun big-time on the "minimize lock hold times" dogma. We haven't shown that this is actually a bottleneck so maintaining atomicity and consistency between message evaluations is much more important. And I'm not even convinced that your micro-optimization is even an optimization. You've set it up so it's going to do a ton more atomic ops and blow the CPU caches with fences in a tight loop.

Reply:
> It doesn't make sense to re-acquire the lock (multiple times?????) on every message. 

That's exactly what `parking_lot` is designed for. 

So yes, acquiring the read lock inside the loop at each iteration is superior, or at least that is what `parking_lot` is designed for. 

The goal is not minimizing the amount of times you (re-)acquire the lock, it's minimizing the chances of a thread being parked(and a thread being parked is particularly bad if that thread is supposed to run non-blocking tasks). 

So by acquiring and releasing the locks as often as possible, you reduce the chance of a thread being parked, since it will have a chance to to a quick acquire/release of the lock in between your own operations.

Reply:
It's not what parking_lot was designed for. `parking_lot` is designed to make this case better, but it doesn't have any goal of forcing you to write code this way. 

The invariant that the state of inner doesn't change in the middle of checking if a single message should be sent to a peer is valuable and it shouldn't be thrown out for micro-optimization.

Reply:
> so maintaining atomicity and consistency between message evaluations is much more important

The question is: why? What explicit logic dictates that for the entire list of messages that are being evaluated, we need `inner` to remain locked by the current thread? Why does the entire loop need to be a critical section?

On the other hand, I can imagine a grandpa voter noting a round in between, which could make several messages irrelevant, at which point we don't want to allow them anymore. Fine-grain locking allows for such "updates" in between evaluation of messages, and I think they're potentially beneficial. 

So I'd say to hold locks for as little as possible, preferably just for a single operation, unless explicit business logic requires a large critical section. If anything, minimizing the scopes of locks held will also make it harder to have unexpected deadlocks.

Reply:
> What explicit logic dictates that for the entire list of messages that are being evaluated, we need inner to remain locked by the current thread.

There isn't any. I think it's reasonable to acquire the lock and hold it for the processing of an entire message, but release afterwards. 

The main thing is that even though the ultra fine-grained acquire-and-release might be technically correct _at the moment_, you are leaving a lot of gaps for subtle races to break stuff in the future. The invariants of the code are much harder to quantify. It may seem like obvious optimization to you, but you need to be conscious that you are making a trade-off against all these things:
  - how long it takes to onboard someone onto this part of the code
  - how much time you have to spend on review
  - how much time it takes to debug

Reply:
Fine, I'll remove those changes, they're not really part of the goal of this PR anyway. 

I do disagree with your arguments, since in my experience it's holding locks for longer than necessary, like all other anti-patterns, that is a subtle source of future problems. 

An example from this very file can be found at https://github.com/paritytech/substrate/pull/2533 and https://github.com/paritytech/substrate/pull/2537